language-plutus-core/examples/Language/PlutusCore/Examples/Data/TreeForest.hs

```diff
@@ -0,0 +1,251 @@
+{-# OPTIONS_GHC -fno-warn-incomplete-uni-patterns #-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+module Language.PlutusCore.Examples.Data.TreeForest
+    ( getBuiltinTree
+    , getBuiltinForest
+    , getBuiltinTreeNode
+    , getBuiltinForestNil
+    , getBuiltinForestCons
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.MkPlc
+import           Language.PlutusCore.Normalize
+
+import           Language.PlutusCore.StdLib.Type
+
+{- Note [Tree]
+Here we encode the following:
+
+    mutual
+      data Tree (A : Set) : Set where
+        node : A -> Forest A -> Tree A
+
+      data Forest (A : Set) : Set where
+        nil  : Forest A
+        cons : Tree A -> Forest A -> Forest A
+
+    mutual
+      mapTree : ∀ A B -> (A -> B) -> Tree A -> Tree B
+      mapTree A B f (node x forest) = node (f x) (mapForest A B f forest)
+
+      mapForest : ∀ A B -> (A -> B) -> Forest A -> Forest B
+      mapForest A B f  nil               = nil
+      mapForest A B f (cons rose forest) = cons (mapTree A B f rose) (mapForest A B f forest)
+
+using this representation:
+
+    {-# OPTIONS --type-in-type #-}
+
+    {-# NO_POSITIVITY_CHECK #-}
+    data IFix {A : Set} (F : (A -> Set) -> A -> Set) (x : A) : Set where
+      wrap : F (IFix F) x -> IFix F x
+
+    Fix₂ : ∀ {A B} -> ((A -> B -> Set) -> A -> B -> Set) -> A -> B -> Set
+    Fix₂ F x y = IFix (λ B P -> P (F λ x y -> B λ R -> R x y)) (λ R -> R x y)
+
+    treeTag : Set -> Set -> Set
+    treeTag T F = T
+
+    forestTag : Set -> Set -> Set
+    forestTag T F = F
+
+    AsTree : (Set -> (Set -> Set -> Set) -> Set) -> Set -> Set
+    AsTree D A = D A treeTag
+
+    AsForest : (Set -> (Set -> Set -> Set) -> Set) -> Set -> Set
+    AsForest D A = D A forestTag
+
+    TreeForest : Set -> (Set -> Set -> Set) -> Set
+    TreeForest =
+      Fix₂ λ Rec A tag ->
+        ∀ R -> tag ((A -> AsForest Rec A -> R) -> R) (R -> (AsTree Rec A -> AsForest Rec A -> R) -> R)
+
+    Tree : Set -> Set
+    Tree = AsTree TreeForest
+
+    Forest : Set -> Set
+    Forest = AsForest TreeForest
+-}
+
+infixr 5 ~~>
```

Comment:
It's certainly useful in this file, because I can actually read

```haskell
star ~~> (star ~~> star ~~> star) ~~> star
```

and

```
[ (vA ~~> forest ~~> vR) ~~> vR
, vR ~~> (tree ~~> forest ~~> vR) ~~> vR
]
```

unlike their non-overloaded versions. On the other hand, I do not seem to need this overloading anywhere else (or it just doesn't work when we need to deal with annotations). I'd prefer to keep it the way it is now, because it's just an example file and I'm not breaking consistency that much, just define one overloaded operator that is handy in this particular file.